### ISSUE TITLE
Rule Change: Relax `no-shadow` again

### ISSUE THREAD

**Issue by Janpot** [eslint/eslint#20978]
### What rule do you want to change?

`no-shadow`

### What change do you want to make?

Generate fewer warnings

### How do you think the change should be implemented?

A new default behavior

### Example code

From MUI/Base UI codebases, we have a pattern that we use all over the place:

```js
const TabList = React.forwardRef(function TabList(props, ref) {
  // ...
});
```

### What does the rule currently do for this code?

Since v10 (#20425 / #20432), both are reported: "'TabList' is already declared in the upper scope". The only options today (`hoist`, `ignoreOnInitialization`, `ignoreTypeValueShadow`, etc.) do not affect this case, so the sole way to silence it is to remove the intentional inner name or add a per-line disable.

### What will the rule do after it's changed?

A named function/class expression assigned (directly or via a wrapper call) to a same-named variable is not reported if the expression's own name binding has zero references within its scope (`variable.references.length === 0`).

Cases that remain reported (the hazard from #20425 is preserved):

```tsx
// Self-reference => `a` inside the body is the UNWRAPPED function, not `wrap(...)`. Still a bug, still reported.
const a = wrap(function a() {
  a(); // referenced -> reported
});
  
// Recursion through the name across a wrapper: fib (inner) !== fib (outer). Still reported.
const fib = memo(function fib(n) {
  return n <= 1 ? 1 : fib(n - 1) + fib(n - 2);
});
```

### Participation

- [x] I am willing to submit a pull request to implement this change.

### AI acknowledgment

- [ ] I did not use AI to generate this issue report.
- [x] (If the above is not checked) I have reviewed the AI-generated content before submitting.

**Comment by DMartens** [comment]
I agree that this should be allowed as long as the variable is not accessed inside the wrapped function.
A PR for this is welcome.
